File: cli/src/huerise_cli/commands/profiles.py
from typing import Annotated
from uuid import UUID

import typer

from huerise_cli.client import api_client
from huerise_cli.config import Config
from huerise_cli.generated.api.alarm_profiles import (
    create_profile as create_profile_api,
)
from huerise_cli.generated.api.alarm_profiles import (
    list_profiles as list_profiles_api,
)
from huerise_cli.generated.models import (
    IntroSchema,
    ProfileCreate,
    ProfileRead,
    RingtoneSchema,
    SunriseSchema,
)
from huerise_cli.output import JsonOption, emit_json, emit_record, emit_table, unwrap

app = typer.Typer(no_args_is_help=True, help="Manage alarm profiles.")


def _profile_row(profile: ProfileRead) -> dict[str, object]:
    return {
        "id": str(profile.id),
        "name": profile.name,
        "default": profile.is_default,
        "scene": profile.sunrise.scene_name,
        "intro_sound": str(profile.intro.sound_id),
        "ringtone_sound": str(profile.ringtone.sound_id),
    }


def _render_profile(profile: ProfileRead, as_json: bool) -> None:
    if as_json:
        emit_json(profile.to_dict())
    else:
        emit_record(_profile_row(profile))


@app.command("list")
def list_profiles(json_: JsonOption = False) -> None:
    """List every alarm profile."""
    with api_client(Config.from_env()) as client:
        profiles = unwrap(list_profiles_api.sync_detailed(client=client))

    if json_:
        emit_json([profile.to_dict() for profile in profiles])
    else:
        emit_table(
            [_profile_row(profile) for profile in profiles],
            empty_message="No profiles yet.",
        )


@app.command("create")
def create_profile(
    name: Annotated[str, typer.Argument(help="Human-readable name for the profile.")],
    intro_sound_id: Annotated[
        UUID, typer.Option(help="Sound id from `huerise sounds list`.")
    ],
    ringtone_sound_id: Annotated[
        UUID, typer.Option(help="Sound id from `huerise sounds list`.")
    ],
    ringtone_volume: Annotated[int, typer.Option(min=0, max=100)] = 80,
    scene_name: Annotated[
        str, typer.Option(help="Hue scene from `huerise rooms get <room>`.")
    ] = "Tageslichtwecker",
    duration_minutes: Annotated[int, typer.Option(min=0, max=120)] = 7,
    brightness_start: Annotated[int, typer.Option(min=1, max=99)] = 1,
    brightness_end: Annotated[int, typer.Option(min=2, max=100)] = 100,
    json_: JsonOption = False,
) -> None:
    """Create a new alarm profile."""
    body = ProfileCreate(
        name=name,
        intro=IntroSchema(sound_id=intro_sound_id),
        ringtone=RingtoneSchema(sound_id=ringtone_sound_id, volume=ringtone_volume),
        sunrise=SunriseSchema(
            scene_name=scene_name,
            duration_minutes=duration_minutes,
            brightness_start=brightness_start,
            brightness_end=brightness_end,
        ),
    )
    with api_client(Config.from_env()) as client:
        profile = unwrap(create_profile_api.sync_detailed(client=client, body=body))
    _render_profile(profile, json_)

File: cli/internal/commands/profiles.go
package commands

import (
	"fmt"
	"strconv"

	"github.com/google/uuid"
	"github.com/mathisarends/huerise/cli/internal/client"
)

type profilesCommand struct {
	List   profilesListCommand   `cmd:"" help:"List every alarm profile."`
	Create profilesCreateCommand `cmd:"" help:"Create a new alarm profile."`
}

type profilesListCommand struct{}

type profilesCreateCommand struct {
	Name            string    `arg:"" help:"Human-readable profile name."`
	IntroSoundID    uuid.UUID `name:"intro-sound-id" required:"" help:"Sound UUID from sounds list."`
	RingtoneSoundID uuid.UUID `name:"ringtone-sound-id" required:"" help:"Sound UUID from sounds list."`
	RingtoneVolume  int       `default:"80" help:"Ringtone volume (0-100)."`
	SceneName       string    `default:"Tageslichtwecker" help:"Hue scene name."`
	DurationMinutes int       `default:"7" help:"Sunrise duration (0-120 minutes)."`
	BrightnessStart int       `default:"1" help:"Starting brightness (1-99)."`
	BrightnessEnd   int       `default:"100" help:"Ending brightness (2-100)."`
}

func (profilesListCommand) Run(runtime *Runtime) error {
	apiClient, err := runtime.client()
	if err != nil {
		return err
	}
	profiles, err := apiClient.ListProfiles(runtime.ctx)
	if err != nil {
		return err
	}
	return runtime.output(profiles, func() error {
		rows := make([][]string, 0, len(profiles))
		for _, profile := range profiles {
			rows = append(rows, profileRow(profile))
		}
		return writeTable(runtime.stdout, []string{"ID", "NAME", "DEFAULT", "SCENE", "INTRO SOUND", "RINGTONE SOUND"}, rows, "No profiles yet.")
	})
}

func (command profilesCreateCommand) Run(runtime *Runtime) error {
	if err := command.validate(); err != nil {
		return err
	}
	apiClient, err := runtime.client()
	if err != nil {
		return err
	}
	request := &client.ProfileCreate{
		Name:  command.Name,
		Intro: client.IntroSchema{SoundID: command.IntroSoundID},
		Ringtone: client.RingtoneSchema{
			SoundID: command.RingtoneSoundID,
			Volume:  client.NewOptInt(command.RingtoneVolume),
		},
		Sunrise: client.NewOptSunriseSchema(client.SunriseSchema{
			SceneName:       client.NewOptString(command.SceneName),
			DurationMinutes: client.NewOptInt(command.DurationMinutes),
			BrightnessStart: client.NewOptInt(command.BrightnessStart),
			BrightnessEnd:   client.NewOptInt(command.BrightnessEnd),
		}),
	}
	response, err := apiClient.CreateProfile(runtime.ctx, request)
	if err != nil {
		return err
	}
	switch result := response.(type) {
	case *client.ProfileRead:
		return writeProfile(runtime, result)
	case *client.HTTPValidationError:
		return validationError(result)
	default:
		return fmt.Errorf("unexpected create profile response %T", response)
	}
}

func (command profilesCreateCommand) validate() error {
	if err := validateVolume(command.RingtoneVolume); err != nil {
		return err
	}
	if command.DurationMinutes < 0 || command.DurationMinutes > 120 {
		return &commandError{Code: "usage", Message: "duration must be between 0 and 120 minutes", ExitCode: 2}
	}
	if command.BrightnessStart < 1 || command.BrightnessStart > 99 {
		return &commandError{Code: "usage", Message: "starting brightness must be between 1 and 99", ExitCode: 2}
	}
	if command.BrightnessEnd < 2 || command.BrightnessEnd > 100 {
		return &commandError{Code: "usage", Message: "ending brightness must be between 2 and 100", ExitCode: 2}
	}
	if command.BrightnessStart >= command.BrightnessEnd {
		return &commandError{Code: "usage", Message: "starting brightness must be less than ending brightness", ExitCode: 2}
	}
	return nil
}

func writeProfile(runtime *Runtime, profile *client.ProfileRead) error {
	return runtime.output(profile, func() error {
		row := profileRow(*profile)
		return writeRecord(runtime.stdout,
			recordField{Name: "id", Value: row[0]},
			recordField{Name: "name", Value: row[1]},
			recordField{Name: "default", Value: row[2]},
			recordField{Name: "scene", Value: row[3]},
			recordField{Name: "intro_sound", Value: row[4]},
			recordField{Name: "ringtone_sound", Value: row[5]},
		)
	})
}

func profileRow(profile client.ProfileRead) []string {
	scene := "-"
	if sunrise, ok := profile.Sunrise.Get(); ok {
		scene = sunrise.SceneName.Or("-")
	}
	return []string{
		profile.ID.String(),
		profile.Name,
		strconv.FormatBool(profile.IsDefault),
		scene,
		profile.Intro.SoundID.String(),
		profile.Ringtone.SoundID.String(),
	}
}
